convex/sources.ts

```diff
@@ -107,130 +70,59 @@ export const addEmbeddingBatch = internalAction(
     ctx,
     { batch }: { batch: { source: Doc<"sources">; texts: string[] }[] }
   ) => {
+    const length = batch.flatMap(({ texts }) => texts.length);
+
     // Calculate all the embeddings for all sources at once.
     const { embeddings, totalTokens, embeddingMs } = await fetchEmbeddingBatch(
       batch.flatMap(({ texts }) => texts)
     );
-    console.log({ batchSize: embeddings.length, totalTokens, embeddingMs });
+
     const offsets = batch.reduce(
       (acc, { texts }) => [...acc, acc[acc.length - 1] + texts.length],
       [0]
     );
-    const vectors = await Promise.all(
-      batch.map(({ source, texts }, idx) => {
-        const vectors = source.chunkIds.map((id, chunkIndex) => ({
-          id,
-          values: embeddings[offsets[idx] + chunkIndex],
-          metadata: { sourceId: source._id, textLen: texts[idx].length },
-        }));
-        return vectors;
+
+    console.log("Check Text", offsets);
+    await Promise.all(
+      batch.map(({ source }, idx) => {
+        source.chunkIds.map(async (id, chunkIndex) => {
+          await ctx.runMutation(internal.sources.chunkPatch, {
+            chunkId: id,
+            embeddings: embeddings[offsets[idx] + chunkIndex],
+          });
+        });
       })
     );
-    await upsertVectors(
-      "chunks",
-      vectors.flatMap((vectors) => vectors)
-    );
-    // The length of all strings put together.
-    const totalLength = batch.reduce(
-      (sum, { texts }) => sum + textLength(texts),
-      0
-    );
+
     await Promise.all(
-      batch.map(async ({ source, texts }) => {
-        const sourceLength = textLength(texts);
-        const portion = sourceLength / totalLength;
+      batch.map(async ({ source }) => {
         await ctx.runMutation(internal.sources.patch, {
           id: source._id,
           patch: {
             saved: true,
-            embeddingMs: Math.ceil(embeddingMs * portion),
-            totalTokens: Math.ceil(totalTokens * portion),
```

Comment:
this might still be useful for folks trying to count tokens to make sure they stay within a max token limit for llm context